### PR TITLE
Disallow fixed-size arrays with zero length

### PR DESCRIPTION
**Pull request by erak**
Fixes https://github.com/ethereum/solidity/issues/5054.

### CODE REVIEW

**Comment by chriseth** [comment]
Perhaps it would be better to disallow zero-size arrays in general at the TypeName level?

**Comment by erak** [comment]
Ok, if we want to disallow them in general, that seems to be the right place.

**Comment by axic** [comment]
See https://github.com/ethereum/solidity/issues/5054#issuecomment-423571627

I think it should be disallowed in every case. Though the ABI encoding allows it.

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 1]
```diff
@@ -241,11 +241,15 @@ void ReferencesResolver::endVisit(ArrayTypeName const& _typeName)
...
 	{
 		TypePointer lengthTypeGeneric = length->annotation().type;
+		auto arrayType = dynamic_cast<ArrayType const*>(baseType.get());
+		bool isByteOrFixedSizeArray = (arrayType && arrayType->isByteArray()) || !baseType->isDynamicallySized();
 		if (!lengthTypeGeneric)
 			lengthTypeGeneric = ConstantEvaluator(m_errorReporter).evaluate(*length);
 		RationalNumberType const* lengthType = dynamic_cast<RationalNumberType const*>(lengthTypeGeneric.get());
 		if (!lengthType || !lengthType->mobileType())
 			fatalTypeError(length->location(), "Invalid array length, expected integer literal or constant expression.");
+		else if (lengthType->isZero() && isByteOrFixedSizeArray)
```
Comment (chriseth) [inline libsolidity/analysis/ReferencesResolver.cpp]: Why did you add the special cases? I think this can be simplified to just `lengthType->isZero()`.

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/ReferencesResolver.cpp (hunk 1 above)
Comment (erak) [reply libsolidity/analysis/ReferencesResolver.cpp]: Ah, of course. I missed that this will only be executed if the length is specified (which is obviously not the case for dynamically-sized arrays 😉 )

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Pleased also update the `Deprecated Elements` section of `050-breaking-changes.rst`.

**Review by chriseth — APPROVED** [review]
